Adds JUJU to the Kubernetes Provider listing

lazypower opened the pull request:
This feature adds Juju provisioning to the kube-up script. It currently
parses out the pre-requisits on debian/ubuntu based systems and installs
them if they are missing.

From there we followed the integration path that was found in the
libvirt-coreos path, implementing the methods found in the boilerplate
and calling juju service calls. There are a few "arbitrary sleeps" in
the code to allow the cloud provider to settle and properly deploy.
These are work-around cases from the script executing faster than juju
was able to communicate from the state server to subsequent nodes. I
left comments inline at these points.

To exercise this:

```
export KUBERNETES_PROVIDER=juju
cluster/kube-up.sh
```

It will spin up a ref arch with 1 Kubernetes Master, 2 minions, and run
the cluster validation checks against the deployment. Bridging the gap
between the juju specific bits and the upstream recommended guides for
getting started with Kubernetes.

To note, if you do not have a "current environment" set in Juju, it will
spin up the quickstart integration wizard in interactive mode, allowing
you to configure juju, and add the proper provider/use it. Otherwise it
assumes you're in the provider you wish to use, and will deploy there.

Commit:
Adds JUJU to the Kubernetes Provider listing

This feature adds Juju provisioning to the kube-up script. It currently
parses out the pre-requisits on debian/ubuntu based systems and installs
them if they are missing.

From there we followed the integration path that was found in the
libvirt-coreos path, implementing the methods found in the boilerplate
and calling juju service calls. There are a few "arbitrary sleeps" in
the code to allow the cloud provider to settle and properly deploy.
These are work-around cases from the script executing faster than juju
was able to communicate from the state server to subsequent nodes. I
left comments inline at these points.

To exercise this:

    export KUBERNETES_PROVIDER=juju
    cluster/kube-up.sh

It will spin up a ref arch with 1 Kubernetes Master, 2 minions, and run
the cluster validation checks against the deployment. Bridging the gap
between the juju specific bits and the upstream recommended guides for
getting started with Juju.

To note, if you do not have a "current environment" set in Juju, it will
spin up the quickstart integration wizard in interactive mode, allowing
you to configure juju, and add the proper provider/use it. Otherwise it
assumes you're in the provider you wish to use, and will deploy there.

erictune commented:
It seems that the flags for the various binaries are stored outside this repo, at places like `https://github.com/whitmo/charm-kubernetes-master/blob/master/files/apiserver.upstart.tmpl`.

For the other cloud providers which support `kube-up.sh`, the flags are in our repository.  This allows us to make changes in a single commit to both the source code and the flags for all those distros.  

Is there a reason to have the code for juju split across repos in this way?

chuckbutler commented:
Thanks for the review!

The current structure of our repositories reflects a pattern we use to separate concerns in the charm store - but this does not reflect the pattern you would need to perform easy updates as the kuberentes project moves foward. I apologize for not recognizing that before issuing the PR.

I'll get the charms/bundle pulled into this branch and we will certainly look into conglomerating our resources that are outside of the GCP namespace so it's easier to track.

filbranden commented on cluster/juju/prereqs/ubuntu-juju.sh:
```diff
@@ -0,0 +1,46 @@
+# Kubernetes Juju Charms project - located here: https://github.com/whitmo/bundle-kubernetes
+
+function check_for_ppa(){
+    grep -s ^ /etc/apt/sources.list /etc/apt/sources.list.d/* | grep juju
```
Why not just:

```shell
  grep -qsw juju /etc/apt/sources.list /etc/apt/sources.list.d/*
```

Use `-q` to make it quiet and `-w` to make sure you're matching a whole word (i.e. you want to match `juju` but not `jujuba`).

filbranden replied:
You want to make the ppa name parameterizable, or rename the function to `check_for_juju_ppa`.

filbranden commented on cluster/juju/prereqs/ubuntu-juju.sh:
```diff
@@ -0,0 +1,46 @@
+}
+
+function package_status(){
+    local pkgstatus=`dpkg-query -W --showformat='${Status}\n' $1|grep "install ok installed"`
```
Please use `$(...)` syntax for command substitution.

There's a problem mixing `local` and `$(...)` in that the `local` will "hide" the error state of the command if it fails. Split it in two lines instead.

You also need double quotes around variables that are supposed to expand to a single name, such as `"$1"`. Probably better to assign that to another variable, e.g. `${pkgname}`, make sure you keep it quoted everywhere since it's always supposed to be a single word.

Instead of using `| grep` I think you can just check that the string matches exactly.

Also, always use `[[ ... ]]` for comparisons since it's a bash built-in with better support for words and nicer matching.

```shell
  local pkgname=$1
  local pkgstatus
  pkgstatus=$(dpkg-query -W --showformat='${Status}\n' "${pkgname}"`
  if [[ "${pkgstatus}" != "install ok installed" ]]; then
    echo "Missing package ${pkgname}"
    sudo apt-get --force-yes --yes install "${pkgname}"
  fi
```

filbranden commented on cluster/juju/prereqs/ubuntu-juju.sh:
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
```
Always use:

```shell
  set -o errexit
  set -o nounset
  set -o pipefail
```

To ensure any errors will not go unnoticed.

You may need to work around some places where you can actually expect failure, e.g. use `|| :` after commands you expect might return non-zero and you're OK with it, and use `${varname-}` in cases where the `varname` variable might not be defined and you're OK with a blank instead.

The point is that it's explicit where you're OK with failures, the default makes it so that failures allow the script to keep running with potential disastrous effects.

filbranden commented on cluster/juju/prereqs/ubuntu-juju.sh:
```diff
@@ -0,0 +1,46 @@
+
+function gather_installation_reqs(){
+
+    ppa_installed=$(check_for_ppa) || ppa_installed=''
```
Use the function as a command in `if` directly, you don't need its output, just it's yes/no status:

```shell
  if check_for_ppa "juju"; then
```

filbranden commented on cluster/juju/prereqs/ubuntu-juju.sh:
```diff
@@ -0,0 +1,46 @@
+    local pkgstatus=`dpkg-query -W --showformat='${Status}\n' $1|grep "install ok installed"`
+    if [ "" == "$pkgstatus" ]; then
+        echo "Missing package $1"
+        sudo apt-get --force-yes --yes install $1
```
Looks a bit too forceful to me... Also, not sure it's OK for us to apt-get on behalf of the user without asking them about it... In any case, I'll let @erictune decide whether that's appropriate here.

erictune replied:
I'm okay with leaving this as is.

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+
```
Same here, use the three `set -o ...` to make sure errors will not go uncaught.

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+    # If something were to happen that I'm not accounting for, do not
+    # punish the user by making them tear things down. In a perfect world
+    # quickstart should handle this situation, so be nice in the meantime
+    local JUJUSTATUS=$(juju status kubernetes-master --format=oneline)
```
local variables are usually lowercase.

Split the `local` and the `status=$(...)` in separate lines.

If `juju` returns zero or non-zero for success/failure, it's better to just use it in the `if`, maybe redirecting stdout or asking it to be quiet:

```shell
  if juju status kubernetes-master --quiet; then
```

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+    local JUJUSTATUS=$(juju status kubernetes-master --format=oneline)
+    if [[ -z "$JUJUSTATUS" ]]; then
+        if [[ -d "~/.juju/current-env" ]]; then
+            juju quickstart -i -e $kube_jujuenv --no-browser -i $kube_bundle_url
```
Double quotes around variables that are supposed to expand to a single token.

We also suggest always using `${...}` for variable names:
https://google-styleguide.googlecode.com/svn/trunk/shell.xml?showone=Variable_expansion#Variable_expansion

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+        if [[ -d "~/.juju/current-env" ]]; then
+            juju quickstart -i -e $kube_jujuenv --no-browser -i $kube_bundle_url
+            # sleeping because of juju bug #
+            sleep 120
```
This is really very long! Can you share the bug number?

This sleep is also common between the if and else, maybe refactor it to after the branch?

chuckbutler replied:
I've tuned this down to 60 seconds, but this can still be racey depending on the provider. I've got a bug filed against this - http://pad.lv/1432759 - and left this update in the code as well for anyone following behind me

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+
+
+function detect-master() {
+    foo=$(juju status --format=oneline kubernetes-master | cut -d' ' -f3)
```
Make this a local variable. Probably deserves a better name...

filbranden replied:
Also, `cut` is usually sensitive to whitespace, e.g. two spaces instead of one makes it break... Prefer using `awk '{print $3}'` to get the third field.

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+
+function detect-master() {
+    foo=$(juju status --format=oneline kubernetes-master | cut -d' ' -f3)
+    export KUBE_MASTER_IP=`echo -n $foo`
```
Why not just this directly? It doesn't look like that `echo` is really doing anything there...

```shell
  export KUBE_MASTER_IP=${foo}
```

Or just bypass the `${foo}` variable and set `KUBE_MASTER_IP` directly. In that case, keep the `export` statement on a separate line to avoid the problem with interaction of `$(...)` and local/export/declare/etc.

erictune commented:
@chuckbutler @whitmo

Have you thought about whether your goal is (1) to have periodic releases of Kubernetes-on-JuJu, or (2) to ensure that Kubernetes-on-JuJu is always compatible with the very latest K8s source code.   Option (2) comes with higher expectations in terms of test automation and your overall level of involvement in the kubernetes project.

If your goal is (1), then we should talk about whether this PR is needed at all.

If it is (2), then this PR is definitely the right direction, but I expect additional steps will be needed too, such as moving JuJu Charm files into the kubernetes repo, as mentioned above; and running kubernets e2e test continuously, etc.

I'm happy to see either outcome.  Also, it is fine to start with 1 and move to 2 in the future.

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+   }
+
+function detect-minions(){
+    KUBE_MINION_IP_ADDRESSES=($(juju run --service kubernetes "unit-get private-address" --format=yaml | grep Stdout | cut -d "'" -f 2))
```
`awk` can do both of what `grep` and `cut` do in a single command... can you use it instead?

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+function sleep-status(){
+    local i=0
+    local maxtime=900
+    local JUJUSTATUS=$(juju status kubernetes-master --format=oneline)
```
lowercase, separate lines for `local` and `...=$(...)`

filbranden replied:
You might want to move the `status=$(juju status ...)` inside the loop and use a `break` to leave it, then you'll have a single call to it in your code, not one before the loop and another at the end of the loop...

Also, if you can do with checking the error status of the command instead then that's probably better (assuming that "started" means zero exit status and anything else means non-zero exit status.)

filbranden commented on cluster/juju/util.sh:
```diff
@@ -0,0 +1,93 @@
+    local JUJUSTATUS=$(juju status kubernetes-master --format=oneline)
+    echo "Waiting up to 15 minutes to allow the cluster to come online... wait for it..."
+
+    while [[ $i < $maxtime ]] && [[ $JUJUSTATUS != *"started"* ]]; do
```
Put the `&&` inside the `[[ ... && ... ]]` expression.

filbranden commented on cluster/validate-cluster.sh:
```diff
@@ -69,7 +69,7 @@ for (( i=0; i<${#MINION_NAMES[@]}; i++)); do
 
     name="${MINION_NAMES[$i]}"
-    if [ "$KUBERNETES_PROVIDER" != "vsphere" ] && [ "$KUBERNETES_PROVIDER" != "vagrant" ] && [ "$KUBERNETES_PROVIDER" != "libvirt-coreos" ]; then
+    if [ "$KUBERNETES_PROVIDER" != "vsphere" ] && [ "$KUBERNETES_PROVIDER" != "vagrant" ] && [ "$KUBERNETES_PROVIDER" != "libvirt-coreos" ] && [ "$KUBERNETES_PROVIDER" != "juju" ]; then
```
I know it's pre-existing, but if possible please convert to use `[[ ... ]]` and to have the `&&` inside the `[[ ... && ... ]]`.

Here and on the other two lines you're updating in this file.

Thanks!

filbranden commented:
Bunch of comments on the bash script... I'll reassign to @erictune to decide whether it's OK to merge though, I'm not very familiar with Juju (well, I know what it is and that's that...) so I'd prefer if someone else would take a look at it.

Let me know if you need me to take another look at the shell scripts later in the code review.

Cheers,
Filipe

chuckbutler commented:
Thank you for the review @filbranden. I'll be working on implementing the comments and should have this ready for re-review on Monday. 

Cheers!

erictune commented:
Thanks, @filbranden.  All your comments look reasonable to me.

Commit:
Implement the feedback from filbranden

- Changed check_for_ppa to be parameterized
- Added bash strictmode
- refactored the package_status method to consume variables and be a bit
  nicer to future re-use of the method.
- Cut out extra echo -n statements in favor of tr -d or native awk
- Refactored branching logic paths to leverage double brackets
- normalized local variable annotation
- Updated globals to be all CAPS
- remainder of filbrandens feedback in validate-cluster.sh

erictune commented:
@chuckbutler hoping to hear your thoughts on https://github.com/GoogleCloudPlatform/kubernetes/pull/5414#issuecomment-79253588

mbruzek commented:
@erictune,

We’ve had a bit of a discussion around this to ensure everyone working on the Juju Kubernetes integration is on the same page. We’re all in agreement that having the Juju Charms and Bundles in the Kubernetes repository is in the best interest of our project moving forward.  We’re actively working on addressing the automation of e2e testing - and currently have some basic stand up testing and validation running in our CI. It’s not as automated as requested above, but we’re working towards that end. If you’re interested in the CI reports, you can view them here:   http://reports.vapour.ws/all-bundle-and-charm-results/gh%3Awhitmo%252Fbundle-kubernetes

Adding the visibility of Juju as an officially supported (and hopefully recommended) method to deploy K8’s - we feel this brings users a new and different way to deploy a cluster that has implications beyond just standing up Kubes. 

We plan to continue adjusting the charms to ensure they work with the latest upstream, and to provide configuration options to run different releases. We would like to continue to support this moving forward as the project evolves.

What do you feel are the next steps for the team to proceed forward?

erictune commented:
I'm happy to hear you plan to work on e2e testing, and I'm happy for users to have Juju as an option for deploying Kubernetes.   

I think the next steps are:
  - you fix the build failure
  - I'll merge this PR
  - over time, the Juju+Kubernetes folks to try to get as many of the e2e tests passing as possible.  
  - as you do that, you will probably send additional PRs to extent the cluster/juju/... files.
  - we may make reasonable requests for features or file reorganizaton on those future PRs

erictune commented:
Looking at the build failure, it looks like it doesn't like the boilerplate at the top of your shell scripts.
You may need to do this to run precommit checks:
``
cd kubernetes/.git/hooks/
ln -s ../../hooks/prepare-commit-msg .
ln -s ../../hooks/commit-msg .
```

erictune commented on cluster/juju/prereqs/ubuntu-juju.sh:
```diff
@@ -0,0 +1,48 @@
+
+set -euo pipefail
+
+# Copyright 2014 Canonical LTD. All rights reserved.
```
I don't think we can take code that has anything other than the Kubernetes boilerplate at the top.  See comment on conversation tab about github hooks.

chuckbutler replied:
Brilliant. Thanks for following up on this Eric. I'll get the headers updated and link in the hooks.
Sent using Boxer
On Mar 26, 2015 1:50 PM, Eric Tune <notifications@github.com> wrote:In cluster/juju/prereqs/ubuntu-juju.sh:
> @@ -0,0 +1,48 @@
> +#!/bin/bash
> +
> +set -euo pipefail
> +
> +# Copyright 2014 Canonical LTD. All rights reserved.

I don't think we can take code that has anything other than the Kubernetes boilerplate at the top.  See comment on conversation tab about github hooks.

—Reply to this email directly or view it on GitHub.

erictune commented:
I've looked several places and I don't see a CLA that would apply to this PR. Let me know if you think this is wrong, or need to talk about it off-PR.

chuckbutler commented:
@erictune we're at a point that we should probably talk off PR. I'm going to pull @whitmo and @mbruzek into the thread.

Commit:
Rewrote shell script headers with accepted kubernetes boilerplate
because of failing build notifications on #5414

chuckbutler commented:
@erictune I've got confirmation from legal that our CLA is now signed, and our team has joined the CLA google group. Can you confirm that this is indeed the case?

erictune commented:
I can't see the CLA yet. Let's wait one day to see if things need time to propagate through my side.

erictune commented:
@chuckbutler contact me at etune@google.com

erictune commented:
Ongoing offline CLA discussion...

erictune commented:
I've verified CLA offline.
